core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Controller.scala

```diff
@@ -64,10 +72,52 @@ class Controller(
     logManager.sendCommitted
   )
 
+  val replayManager = new GlobalReplayManager(
+    () => {
+      cp.asyncRPCClient.sendToClient(WorkflowRecoveryStatus(true))
+    },
+    () => {
+      cp.asyncRPCClient.sendToClient(WorkflowRecoveryStatus(false))
+    }
+  )
+
+  def setupReplay(): Unit = {
+    if (controllerConfig.replayTo.isDefined) {
+      replayManager.markRecoveryStatus(CONTROLLER, isRecovering = true)
+
+      val (processSteps, messages) = ReplayLogGenerator.generate(logStorage)
```

Comment:
I wonder if is a good idea for this method/object to be merged into `logManager`?

Reply:
logManager specifies the writing destination, while the log records may be retrieved from another source which is different from the writing destination.